text/0000-transaction-refactoring.md

```diff
@@ -0,0 +1,327 @@
+- Proposal Name: Batched Transactions
+- Start Date: 2019-07-22
+- NEP PR: [nearprotocol/neps#0000](https://github.com/nearprotocol/neps/pull/0000)
+- Issue(s): 
+
+# Summary
+[summary]: #summary
+
+Refactor signed transactions and receipts to support batched atomic transactions and data dependency.
+
+# Motivation
+[motivation]: #motivation
+
+It simplifies account creation, by supporting batching of multiple transactions together instead of
+creating more complicated transaction types.
+
+For example, we want to create a new account with some account balance and one or many access keys, deploy a contract code on it and run an initialization method to restrict access keys permissions for a `proxy` function.
+
+To be able to do this now, we need to have a `CreateAccount` transaction with all the parameters of a new account.
+Then we need to handle it in one operation in a runtime code, which might have duplicated code for executing some WASM code with the rollback conditions.
+
+Alternative to this is to execute multiple simple transactions in a batch within the same block.
+It has to be done in a raw without any commits to the state until the entire batch is completed.
+We propose to support this type of transaction batching to simplify the runtime.
+
+This change
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+### New transaction and receipts
+
+Previously, in the runtime to produce a block we first executed new signed transactions and then executed received receipts. It resulted in duplicated code that might be shared across similar actions, e.g. function calls for async calls, callbacks and self-calls.
+It also increased the complexity of the runtime implementation.
+
+This NEP proposes changing it by first converting all signed transactions into receipts and then either execute them immediately before received receipts, or put them into the list of the new receipts to be routed.
+To achieve this, NEP introduces a new message `Action` that represents one of atomic actions, e.g. a function call.
+`TransactionBody` is now called just `Transaction`. It contains the list of actions that needs to be performed in a single batch and the information shared across these actions.
+
+`Transaction` contains the following fields
+- `originator_id` is an account ID of the transaction originator.
+- `public_key` is to identify the access key used to signs the transaction.
+- `nonce` is used to deduplicate and order transactions.
+- `receiver_id` is where the transaction has to be routed.
+- `action` is the list of actions to perform.
+
+An `Action` can be of the following:
+- `CreateAccount` creates a new account. It has to be the first action, and the receiver has to be a new account. The action will fail if the account already exists.
```

Comment:
Why does it *have* to be the first action?

Reply:
Until we have the account deletion, it doesn't make sense for trying to create an account after some other action.